### ISSUE TITLE
Only compute per_sample_gradients for parameters that are optimized

### ISSUE THREAD

**Issue by timudk** [pytorch/opacus#501]
## 🚀 Feature
Currently, per_sample_gradients are computed for each parameter that is part of the backward pass. I am requesting a feature that only computes per_sample_gradients for parameters that are actually optimized.

## Motivation
The technique is, for example, used in this paper: https://arxiv.org/abs/2110.06500. Using this technique, the authors were able to fine-tune a billion parameter model with DPSGD. If they computed per_sample_gradients for all parameters that are part of the backward pass, this would have never been possible since per_sample_gradients require a very large amount of memory.

**Comment by alexandresablayrolles** [comment]
If I am not mistaken, the GradSampleModule only adds hooks to parameters that have `require_grad=True`, which means that the per_sample_gradients are only computed for parameters that are optimized.

**Comment by timudk** [comment]
Not necessarily. When fine-tuning a particular layer of a network, we need to set require_grad=True for all layers afterwards (since we need those gradients to backpropagate to the layer we optimize), even though they are not optimized.

**Comment by alexandresablayrolles** [comment]
I don't think this is correct. I think if you only want to fine-tune a particular layer, you need to set require_grad=True for this layer *only*. Pytorch will still figure out which layer to backprop through. 
Let me know if that works for you, and feel free to close to issue if it does.
